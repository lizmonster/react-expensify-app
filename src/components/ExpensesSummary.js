// Rendered above filter
// expenseCount prop, expensesTotal prop - connect to store
// both selectors needed

// Tests
// One expense or multiple expenses or none
// Format expense total using numeraljs

// Display should change with filters

import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ numberOfExpenses, totalExpenses }) => {
    
    const expensesWord = numberOfExpenses > 1 ? 'expenses' : 'expense';
    const totalExpensesFormatted = numeral(totalExpenses / 100).format('$0,0.00');

    return (
        <div>
            {
                numberOfExpenses === 0 ? (
                    <h1>Viewing 0 expenses</h1>
                ) :
                    (
                        <h1>Viewing { numberOfExpenses } { expensesWord } totalling { totalExpensesFormatted }</h1>
                    )
            }

        </div>
    )
};

const mapStateToProps = (state) => {
    const allExpenses = selectExpenses(state.expenses, state.filters);

    return {
        numberOfExpenses: allExpenses.length,
        totalExpenses: expensesTotal(allExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
