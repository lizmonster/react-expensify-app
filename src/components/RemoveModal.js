import React from 'react';
import Modal from 'react-modal';

const OptionModal = (props) => (
    <Modal
        isOpen={!!props.removingExpense}
        ariaHideApp={false}
        contentLabel="Remove Expense"
        closeTimeoutMS={
            200
        }
        className="modal"
    >
        <h3 className="modal__title">Remove Expense</h3>
        <h3 className="modal__name">"{props.expenseName}" ?</h3>
        <button className="button modal__button" onClick={props.handleRemove}>Yes</button>
        <button className="button modal__button" onClick={props.onDismiss}>No</button>
    </Modal>
);

export default OptionModal;