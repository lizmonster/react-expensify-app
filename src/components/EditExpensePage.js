import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import RemoveModal from './RemoveModal';

import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    state = {
        removingExpense: undefined
    }
    handleRemoveExpense = (expense) => {
        this.setState(() => ({ removingExpense: true }));
    }
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/dashboard');
    };
    onRemove = () => {
        this.setState(() => ({
            removingExpense: undefined
        }));
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/dashboard');
    };
    onDismiss = () => {
        this.setState(() => ({
            removingExpense: undefined
        }));
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.handleRemoveExpense}>Remove Expense</button>
                </div>
                <RemoveModal 
                    expenseName={this.props.expense.description}
                    removingExpense={this.state.removingExpense}
                    handleRemove={this.onRemove}
                    onDismiss={this.onDismiss} 
                />
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id),
        removingExpense: state.removingExpense
    };
};

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: ({ id }) => dispatch(startRemoveExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
