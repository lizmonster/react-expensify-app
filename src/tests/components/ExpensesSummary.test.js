import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render empty ExpensesSummary correctly', () => {
    const wrapper = shallow(<ExpensesSummary numberOfExpenses={0} totalExpenses={0} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render a single expense correctly', () => {
    const wrapper = shallow(<ExpensesSummary numberOfExpenses={1} totalExpenses={115} />)
    expect(wrapper).toMatchSnapshot();
});

test('should render multiple expenses correctly', () => {
    const wrapper = shallow(<ExpensesSummary numberOfExpenses={5} totalExpenses={932} />)
    expect(wrapper).toMatchSnapshot();
});
