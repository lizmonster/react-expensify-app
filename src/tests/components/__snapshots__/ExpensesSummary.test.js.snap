// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a single expense correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $1.15
  </p>
</div>
`;

exports[`should render empty ExpensesSummary correctly 1`] = `
<div>
  <p>
    Viewing 0 expenses
  </p>
</div>
`;

exports[`should render multiple expenses correctly 1`] = `
<div>
  <p>
    Viewing 
    5
     
    expenses
     totalling 
    $9.32
  </p>
</div>
`;
